src/library/scala/collection/MutationTracker.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.collection
+
+import java.util.ConcurrentModificationException
+
+/**
+ * A utility that tracks mutations; a client is meant to
+ * call [[addMutation()]] each time it mutates itself, and
+ * to have its iterator check that the client hasn't mutated
+ * itself since the iterator's creation either by wrapping the
+ * iterator using [[checkedIterator]] or by manually using a
+ * [[checker]].
+ */
+private final class MutationTracker private {
+  private[this] var mutationCount: Int = 0
+
+  def addMutation(): Unit = mutationCount += 1
+
+  def checker: Checker = new Checker
+
+  def checkedIterator[A](it: Iterator[A]): Iterator[A] = new CheckedIterator(it)
+
+  /**
+   * A utility that can be used to check that its creator
+   * has not mutated itself since this checker's creation.
+   */
+  final class Checker private[MutationTracker] {
+    private[this] val expectedCount = mutationCount
+
+    /**
+     * Checks whether or not this checker's creator has mutated
+     * itself since this checker's creation, throwing an exception,
+     * if it has.
+     *
+     * @param message the exception message in case of mutations
+     * @throws ConcurrentModificationException if this checker's creator
+     *                                         has mutated itself
+     */
+    @throws[ConcurrentModificationException]
+    def checkMutations(message: String): Unit = {
+      if (mutationCount != expectedCount) throw new ConcurrentModificationException(message)
+    }
+
+    /**
+     * Checks whether or not this checker's creator has mutated
+     * itself since this checker's creation, throwing an exception,
+     * if it has. This method produces an exception message saying
+     * that it was called by an iterator whose backing collection
+     * was mutated after the iterator's creation.
+     *
+     * @throws ConcurrentModificationException if the collection has
+     *                                         been mutated
+     */
+    @throws[ConcurrentModificationException]
+    def checkMutationsForIterator(): Unit =
+      checkMutations("mutation occurred after iterator creation")
+  }
+
+  private final class CheckedIterator[A](underlying: Iterator[A]) extends AbstractIterator[A] {
+    private[this] val checker = new Checker
+
+    def hasNext: Boolean = underlying.hasNext
```

Comment:
Due to the contract of iterators, `hasNext` often actually walks the internal data structure of the collection, and is therefore also at risk of causing weird bugs like NPEs in case of concurrent access. Therefore, it should also be protected `checkMutationsForIterator()`. And yes, that's worse for performance.

Reply:
That's not needed for the purpose of best-effort fast-failing though. As I understand it, the goal of this change is to prevent the user from iterating over a modified collection if we can. Emitting an error on getting the next element will accomplish that, even if hasNext throws or returns a wrong result.

Reply:
do you think we could get away with just checking in `hasNext` either? I'm guessing not...

Reply:
> That's not needed for the purpose of best-effort fast-failing though.

given that, I think checking for `hasNext` and not `next()` is probably safe enough, and safer than just checking in `next()`, since `next()` is typically only called after `hasNext`